### ISSUE TITLE
`Identity` and `Recipient` Not Available at Runtime

### ISSUE THREAD

**Issue by Alchemyst0x** [woodruffw/pyrage#81]
First off, thanks for this package, I appreciate having these bindings available.

I'm encountering an issue where the `Identity` and `Recipient` types are not available at runtime, even though they are defined in the source code.

### Steps to Reproduce

1. Install the `pyrage` and `pyrage-stubs` packages.
2. Try to import `Identity` and `Recipient`:
   ```python
   from pyrage import Identity, Recipient
   ```

3. This results in the following error:
   ```python
   ImportError: cannot import name 'Identity' from 'pyrage'
   ```

### What I've Tried

- I've checked the installed package, and it seems that `Identity` and `Recipient` are not being exposed in the `__init__.py` file.
- The type checker (e.g., in VSCode with Pyright) can see these types, but they are not available at runtime.
- I've tried modifying the `pyrage-stubs` package (e.g., adding `py.typed`, converting `*.py` files to `*.pyi`, adding `__all__`), but the issue persists.
- I've also tried rebuilding the package using `maturin` and even starting fresh with a new project structure using [rye](https://rye.astral.sh/guide/rust/), but no luck.

### Additional Context

- I'm working on an Apple M1 machine, but I don't think this issue is platform-specific.
- I've pushed a fork of the repository with my attempts to resolve the issue: [here](https://github.com/Alchemyst0x/pyrage.git)
- I have a temporary workaround in my project where I define the following:

  ```python
  from pyrage import ssh, x25519
  
  Identity = ssh.Identity | x25519.Identity
  Recipient = ssh.Recipient | x25519.Recipient
  
  type IdentityT = ssh.Identity | x25519.Identity
  type RecipientT = ssh.Recipient | x25519.Recipient
  ```

  Doing this allows for using the types as both annotations and for instance checking them, e.g.,
  
  ```python
  from collections.abc import Sequence
  from typing import Any, TypeGuard
  
  from pyrage import IdentityError, RecipientError, x25519
  from .types.age import Identity, IdentityT, Recipient, RecipientT
  
  
  def _is_sequence(obj: Any) -> TypeGuard[Sequence]:
      return isinstance(obj, Sequence) and not isinstance(obj, str | bytes | bytearray)
  
  
  def ensure_recipient_sequence(obj: Any) -> Sequence[RecipientT]:
      try:
          if isinstance(obj, str) and (_recipient := x25519.Recipient.from_str(obj)):
              obj = _recipient
          if isinstance(obj, Recipient) and not _is_sequence(obj):
              return [obj]
          if _is_sequence(obj) and all(isinstance(_, Recipient) for _ in obj):
              return obj
          raise RecipientError
      except Exception as exc:
          raise RecipientError from exc
  ```

I am not very familiar with `maturin` or Rust development in general, so any help or guidance on how to properly expose `Identity` and `Recipient` at runtime would be greatly appreciated! My best guess is that while these are certainly type annotations, they may just need to be defined as such `Union` objects in the Rust source.

**Comment by woodruffw** [comment]
Thanks for the report.

Can you say more about what you're trying to accomplish? `pyrage` is the concrete implementation and `pyrage-stubs` is the PEP 561 type information package, meaning that you should install both into your development environment if you want type checking. 

Notably, you should **not** import `pyrage-stubs` (or modify it on disk): it's intended only for the typechecker, not for API usage. 

TL;DR: your custom unions/type aliases look fine. `pyrage` doesn't expose an `Identity` union type in its actual code, so defining it yourself is perfectly appropriate.

(We _could_ expose one, although that would require converting `pyrage` from a pure-pyO3 package into a hybrid one with Python source files.)

**Comment by Alchemyst0x** [comment]
The reason I was attempting to do things this way to begin with is due to the fact that, if you have the stub package installed as I do, and you are importing directly from `pyrage`, users are lead to believe that there are in fact such objects for `Identity` and `Recipient`. 

I've attached a screenshot of what this looks like in VSCode with Pylance, to show what I mean. This is using the latest release without modifying anything. If I run this code, I will get an import error immediately due to the fact that these types are defined, but as you've stated, they are not actually exported from the core package.


![2024 10 01T16 52 37-tVGNbM0o](https://github.com/user-attachments/assets/6b08ca03-ff25-470e-9215-5b1d17d8e7aa)

Meanwhile, somewhat ironically, as shown here for `x25519.Identity.generate()`, I am able to fully utilize this function as intended and generate an `Identity`, but the type checker is unable to infer anything about this function or its return types.  

![2024 10 01T17 06 30-pGaG5Ko3](https://github.com/user-attachments/assets/5710d7f7-6472-49d5-8e2f-56995af31788)

So to answer your question, my hope was that I could correct this inconsistency, otherwise I would not have been modifying anything. Since I could not, I just wanted to bring it to your attention.

**Comment by woodruffw** [comment]
> I've attached a screenshot of what this looks like in VSCode with Pylance, to show what I mean. This is using the latest release without modifying anything. If I run this code, I will get an import error immediately due to the fact that these types are defined, but as you've stated, they are not actually exported from the core package.

Ah, gotcha -- I wasn't aware that it an IDE would present those aliases as if they're actually importable. That's helpful context; I'll see if I can improve the hinting there tonight or tomorrow.

> Meanwhile, somewhat ironically, as shown here for `x25519.Identity.generate()`, I am able to fully utilize this function as intended and generate an `Identity`, but the type checker is unable to infer anything about this function or its return types.

Not sure what's up with that one -- from the stubs, it's definitely annotated to return an `x25519.Identity` there:

https://github.com/woodruffw/pyrage/blob/cd023bca1a89d2a6b76cd5f0e7dc323b4f3a5424/pyrage-stubs/pyrage-stubs/x25519.py#L5-L7

I don't use Pylance, however, so I don't have a clear sense of how it differs from mypy.

**Comment by Alchemyst0x** [comment]
> Not sure what's up with that one -- from the stubs, it's definitely annotated to return an `x25519.Identity` there:
> 
> https://github.com/woodruffw/pyrage/blob/cd023bca1a89d2a6b76cd5f0e7dc323b4f3a5424/pyrage-stubs/pyrage-stubs/x25519.py#L5-L7

> I don't use Pylance, however, so I don't have a clear sense of how it differs from mypy.

I will take another look at this when I get a chance here shortly and see if I can get things to look right on my end, since we are using different environments. As far as I know, for VSCode, Pylance is the default for Python; it's just [Pyright](https://github.com/microsoft/pyright).

Regardless, I actually ran into another peculiarity here, and when this happened I started to wonder if this might be specific to my environment, so I actually cleared all the cached packages and other Python files out, syncing everything anew just to confirm. It does not appear to be anything on my side specifically. Rather than get into every little detail, let me just ask:

https://github.com/woodruffw/pyrage/blob/cd023bca1a89d2a6b76cd5f0e7dc323b4f3a5424/pyrage-stubs/pyrage-stubs/__init__.pyi#L19

Is this meant to be typed for `in_io`, `out_io`, and `Recipients` parameters? For me, this function will only work if provided all three arguments. Meanwhile, Pylance is again alerting to the fact that this, according to the typing, is incorrect.

**Comment by woodruffw** [comment]
> Rather than get into every little detail, let me just ask:
> 
> https://github.com/woodruffw/pyrage/blob/cd023bca1a89d2a6b76cd5f0e7dc323b4f3a5424/pyrage-stubs/pyrage-stubs/__init__.pyi#L19
> 
> Is this meant to be typed for `in_io`, `out_io`, and `Recipients` parameters? For me, this function will only work if provided all three arguments. Meanwhile, Pylance is again alerting to the fact that this, according to the typing, is incorrect.

Yeah, that one looks like a typing bug, introduced with streaming in #77.

The actual signature is here:

https://github.com/woodruffw/pyrage/blob/cd023bca1a89d2a6b76cd5f0e7dc323b4f3a5424/src/lib.rs#L268-L273

If you'd like to send a PR to fix that, I'd be happy to merge it and get a fixed stubs package out ASAP. Otherwise I'll have time to fix it when I look at this more generally.

**Comment by woodruffw** [comment]
(More generally, this suggests that the CI for `pyrage` should probably be running `mypy` on the unit tests...)

**Comment by Alchemyst0x** [comment]
Got it, opened a PR and I believe I have managed to address the other issues as well. Not sure if you will agree about making the Identity and Recipient types private, but this seemed to be a reasonable solution for the time, as I understood this to mean that these were not originally intended to be part of the public API to begin with.

**Comment by woodruffw** [comment]
> Not sure if you will agree about making the Identity and Recipient types private, but this seemed to be a reasonable solution for the time, as I understood this to mean that these were not originally intended to be part of the public API to begin with.

Makes sense to me -- those were indeed intended to be private, in the sense that they were there just to provide uniform typing within the stubs themselves. From re-reading the typing docs, it seems like prefixing them with an `_` is absolutely correct for that purpose.

**Comment by woodruffw** [comment]
Thanks for the PR -- I've cut 1.2.2 of the stubs and they should be available on PyPI shortly. Let me know if you have any other typing issues (and sorry for the misunderstanding originally).
